File: src/agent_hub_bridges/_common/journal.py
"""Outgoing message journal for crash-safe bridge sends (issue #183 / agent-hub#168).

ブリッジが ``hub.send()`` を呼ぶ前に journal entry を書き出し、
ack 受信後に削除する。再起動時は pending entry を replay することで
「bridge クラッシュによる outgoing メッセージ消失」を防止する。

設計ポイント
------------
- **at-least-once**: journal に書いてから送信するため、クラッシュ後に
  再送されると重複する可能性がある。replay 時の重複防止 (idempotency) は
  TODO (Issue #183) — operator の Go 待ち。

- **フォーマット**: JSONL (1 行 1 エントリ)。末尾 append + ``os.fsync`` で
  crash-safe に書き出し、削除時は tmpfile + ``os.replace`` (atomic rename) で
  安全に更新する。

- **保存先**: デフォルト ``~/.agent-hub/journals/<name>.journal``。
  環境変数 ``AGENT_HUB_JOURNAL_DIR`` でディレクトリを上書き可能。

NOTE: 冪等性 (idempotency_key) 対応は TODO (issue #183)。
  - Option A: ``send_message`` MCP ツールに ``idempotency_key`` パラメータを追加し
    server 側で重複チェック (agent-hub L1 変更が必要)。
  - Option B: journal entry に ``sent_message_id`` を記録し bridge 側で照合。
  operator の Go 取得後に別 PR で実装する。
"""

from __future__ import annotations

import contextlib
import json
import logging
import os
import uuid
from dataclasses import asdict, dataclass, field
from datetime import UTC, datetime
from pathlib import Path

logger = logging.getLogger(__name__)

_JOURNAL_DIR_ENV = "AGENT_HUB_JOURNAL_DIR"
_DEFAULT_JOURNAL_DIR = Path.home() / ".agent-hub" / "journals"


def journal_dir() -> Path:
    """journal ディレクトリの Path を返す。

    ``AGENT_HUB_JOURNAL_DIR`` 環境変数があればそれを優先。
    なければ ``~/.agent-hub/journals/``。
    """
    env_val = os.environ.get(_JOURNAL_DIR_ENV)
    return Path(env_val) if env_val else _DEFAULT_JOURNAL_DIR


@dataclass
class JournalEntry:
    """journal の 1 エントリ。``hub.send()`` の引数と同じ形式。

    Attributes:
        id: エントリの UUID。将来 idempotency_key として server に渡す予定
            (TODO: issue #183)。
        to: 送信先 handle (例: ``"@planner"``)。
        message: 送信するメッセージ本文。
        caused_by: 因果チェーン用の元 message ID (issue #162 / agent-hub)。省略可。
        created_at: 書き込み時刻 (ISO-8601 UTC)。
    """

    id: str
    to: str
    message: str
    caused_by: str | None = None
    created_at: str = field(default_factory=lambda: datetime.now(UTC).isoformat())


class Journal:
    """JSONL 形式の outgoing message journal。

    スレッドセーフ・crash-safe:

    - ``write()``: 末尾 append + ``os.fsync`` で即時永続化。
    - ``delete()``: tmpfile + ``os.replace`` (atomic rename) で安全に更新。
    - ``load_all()``: 破損行はスキップして warning ログのみ (クラッシュしない)。

    使い方::

        journal = Journal("claude-impl")
        entry = journal.make_entry(to="@alice", message="hello", caused_by=msg.id)
        journal.write(entry)
        try:
            await hub.send(to=entry.to, message=entry.message, caused_by=entry.caused_by)
            journal.delete(entry.id)
        except Exception:
            # 失敗時は entry を残す → 次回起動時に replay
            raise
    """

    def __init__(self, name: str, base_dir: Path | None = None) -> None:
        """
        Args:
            name: ジャーナルファイル名のベース (拡張子 ``.journal`` が付く)。
                  通常はブリッジの user handle (例: ``"claude-impl"``)。
            base_dir: ジャーナルディレクトリ。``None`` なら :func:`journal_dir` を使う。
        """
        base = base_dir if base_dir is not None else journal_dir()
        self._path = base / f"{name}.journal"

    @property
    def path(self) -> Path:
        """journal file の Path。"""
        return self._path

    def make_entry(
        self,
        *,
        to: str,
        message: str,
        caused_by: str | None = None,
    ) -> JournalEntry:
        """新規 :class:`JournalEntry` を生成する (書き込みは行わない)。"""
        return JournalEntry(
            id=str(uuid.uuid4()),
            to=to,
            message=message,
            caused_by=caused_by,
        )

    def write(self, entry: JournalEntry) -> bool:
        """journal に entry を末尾 append する。

        ``os.fsync`` で即時ディスク反映。

        Returns:
            True  — 書き込み成功。
            False — 書き込み失敗 (warning ログのみ、例外は上げない)。
                    呼び出し側は False を受け取ったら送信を中止すること。
        """
        try:
            self._path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
            with open(self._path, "a", encoding="utf-8") as f:
                f.write(json.dumps(asdict(entry), ensure_ascii=False) + "\n")
                f.flush()
                os.fsync(f.fileno())
            logger.debug("Journal write: entry %s → %s", entry.id, self._path)
            return True
        except Exception:
            logger.warning(
                "Failed to write journal entry %s to %s — send will be aborted",
                entry.id,
                self._path,
                exc_info=True,
            )
            return False

    def delete(self, entry_id: str) -> None:
        """entry_id に一致するエントリを journal から削除する。

        tmpfile + ``os.replace`` (atomic rename) で crash-safe に更新する。
        失敗しても例外を上げない (warning ログのみ)。
        """
        if not self._path.exists():
            return
        entries = self.load_all()
        remaining = [e for e in entries if e.id != entry_id]
        if len(remaining) == len(entries):
            logger.debug("Journal delete: entry %s not found (already deleted?)", entry_id)
            return
        self._write_all(remaining)
        logger.debug("Journal delete: entry %s removed from %s", entry_id, self._path)

    def load_all(self) -> list[JournalEntry]:
        """journal の全エントリを読み込む。

        破損行はスキップして warning ログのみ (クラッシュしない)。
        ファイルが存在しない場合は空リストを返す。
        """
        if not self._path.exists():
            return []
        entries: list[JournalEntry] = []
        try:
            with open(self._path, encoding="utf-8") as f:
                for lineno, raw in enumerate(f, 1):
                    raw = raw.strip()
                    if not raw:
                        continue
                    try:
                        data = json.loads(raw)
                        # 前方互換: 未知フィールド (将来バージョンで追加) を無視する。
                        # JournalEntry.__dataclass_fields__ で既知フィールドをフィルタし
                        # TypeError を防ぐ (issue #183 reviewer Minor 2)。
                        known = JournalEntry.__dataclass_fields__.keys()
                        entries.append(JournalEntry(**{k: v for k, v in data.items() if k in known}))
                    except (json.JSONDecodeError, TypeError) as exc:
                        logger.warning(
                            "Skipping corrupt journal line %d in %s: %s",
                            lineno,
                            self._path,
                            exc,
                        )
        except Exception:
            logger.warning(
                "Failed to read journal from %s — treating as empty",
                self._path,
                exc_info=True,
            )
        return entries

    def _write_all(self, entries: list[JournalEntry]) -> None:
        """journal を entries で上書きする (atomic rename)。

        entries が空の場合はファイルを削除する。
        失敗しても例外を上げない (warning ログのみ)。
        """
        if not entries:
            with contextlib.suppress(FileNotFoundError):
                self._path.unlink()
            logger.debug("Journal cleared (empty), file removed: %s", self._path)
            return

        # PID を含めることでマルチプロセス環境での tmp ファイル衝突を防ぐ
        # (reviewer Minor 1: issue #183)
        tmp_path = self._path.with_name(f"{self._path.stem}.{os.getpid()}.journal.tmp")
        try:
            with open(tmp_path, "w", encoding="utf-8") as f:
                for e in entries:
                    f.write(json.dumps(asdict(e), ensure_ascii=False) + "\n")
                f.flush()
                os.fsync(f.fileno())
            os.replace(tmp_path, self._path)
        except Exception:
            logger.warning(
                "Failed to rewrite journal at %s",
                self._path,
                exc_info=True,
            )
            with contextlib.suppress(FileNotFoundError):
                tmp_path.unlink()

File: bridge-claude2/cmd/bridge/journal.go
// journal.go — Outgoing message journal (Python: _common/journal.py の直訳)
//
// bridge が hub.SendMessage() を呼ぶ前に journal entry を書き出し、
// 送信完了後に削除する。再起動時は pending entry を replay することで
// 「bridge クラッシュによる outgoing メッセージ消失」を防止する。
//
// at-least-once セマンティクス: journal に書いてから送信するため、
// クラッシュ後に再送されると重複する可能性がある (idempotency は TODO)。
//
// フォーマット: JSONL (1 行 1 エントリ)。
// 保存先: ~/.agent-hub/journals/<user>.journal (env AGENT_HUB_JOURNAL_DIR で上書き可)。
package main

import (
	"bufio"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sync"
	"time"
)

const journalDirEnv = "AGENT_HUB_JOURNAL_DIR"

// journalEntry は journal の 1 エントリ。hub.SendMessage の引数と同じ形式。
type journalEntry struct {
	ID        string `json:"id"`
	To        string `json:"to"`
	Message   string `json:"message"`
	CausedBy  string `json:"caused_by,omitempty"`
	CreatedAt string `json:"created_at"`
}

// Journal は JSONL 形式の outgoing message journal。
// スレッドセーフ: mu で write/delete/loadAll を保護する。
type Journal struct {
	mu   sync.Mutex
	path string
}

// newJournal は指定 user の Journal を生成する。
func newJournal(user string) *Journal {
	dir := os.Getenv(journalDirEnv)
	if dir == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			home = "/tmp"
		}
		dir = filepath.Join(home, ".agent-hub", "journals")
	}
	return &Journal{path: filepath.Join(dir, user+".journal")}
}

// makeEntry は新規 journalEntry を生成する (書き込みは行わない)。
func (j *Journal) makeEntry(to, message, causedBy string) journalEntry {
	return journalEntry{
		ID:        newRandomID(),
		To:        to,
		Message:   message,
		CausedBy:  causedBy,
		CreatedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}
}

// newRandomID は crypto/rand を使った UUID-like なランダム ID を生成する。
func newRandomID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		// フォールバック: 時刻ベース (本番では発生しない想定)
		return fmt.Sprintf("%x", time.Now().UnixNano())
	}
	// UUID v4 形式 (xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx)
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%s-%s-%s-%s-%s",
		hex.EncodeToString(b[0:4]),
		hex.EncodeToString(b[4:6]),
		hex.EncodeToString(b[6:8]),
		hex.EncodeToString(b[8:10]),
		hex.EncodeToString(b[10:16]),
	)
}

// write は journal に entry を末尾 append する。
// Sync で即時ディスク反映。失敗時は false を返す (例外は上げない)。
// 呼び出し側は false を受け取ったら送信を中止すること。
func (j *Journal) write(entry journalEntry) bool {
	j.mu.Lock()
	defer j.mu.Unlock()

	if err := os.MkdirAll(filepath.Dir(j.path), 0o700); err != nil {
		slog.Warn("journal: failed to create journal dir", "path", j.path, "err", err)
		return false
	}
	f, err := os.OpenFile(j.path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o600)
	if err != nil {
		slog.Warn("journal: failed to open journal file", "path", j.path, "err", err)
		return false
	}
	defer f.Close()

	data, _ := json.Marshal(entry)
	if _, err := fmt.Fprintf(f, "%s\n", data); err != nil {
		slog.Warn("journal: failed to write entry", "id", entry.ID, "err", err)
		return false
	}
	if err := f.Sync(); err != nil {
		slog.Warn("journal: fsync failed", "id", entry.ID, "err", err)
		return false
	}
	slog.Debug("journal: write", "id", entry.ID, "path", j.path)
	return true
}

// loadAll は journal の全エントリを読み込む。
// 破損行はスキップして WARNING ログのみ (bridge を落とさない)。
// ファイルが存在しない場合は nil を返す。
func (j *Journal) loadAll() []journalEntry {
	j.mu.Lock()
	defer j.mu.Unlock()
	return j.loadAllLocked()
}

func (j *Journal) loadAllLocked() []journalEntry {
	f, err := os.Open(j.path)
	if err != nil {
		if !os.IsNotExist(err) {
			slog.Warn("journal: failed to open for read", "path", j.path, "err", err)
		}
		return nil
	}
	defer f.Close()

	var entries []journalEntry
	scanner := bufio.NewScanner(f)
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		raw := scanner.Bytes()
		if len(raw) == 0 {
			continue
		}
		var e journalEntry
		if err := json.Unmarshal(raw, &e); err != nil {
			slog.Warn("journal: skipping corrupt line", "line", lineNo, "path", j.path, "err", err)
			continue
		}
		entries = append(entries, e)
	}
	if err := scanner.Err(); err != nil {
		slog.Warn("journal: scan error", "path", j.path, "err", err)
	}
	return entries
}

// delete は entry_id に一致するエントリを journal から削除する。
// tmpfile + atomic rename で crash-safe に更新する。
// 失敗しても例外を上げない (WARNING ログのみ)。
func (j *Journal) delete(entryID string) {
	j.mu.Lock()
	defer j.mu.Unlock()

	entries := j.loadAllLocked()
	remaining := make([]journalEntry, 0, len(entries))
	found := false
	for _, e := range entries {
		if e.ID == entryID {
			found = true
			continue
		}
		remaining = append(remaining, e)
	}
	if !found {
		slog.Debug("journal: entry not found (already deleted?)", "id", entryID)
		return
	}
	j.writeAllLocked(remaining)
	slog.Debug("journal: deleted entry", "id", entryID, "path", j.path)
}

// writeAllLocked は journal を remaining で上書きする (atomic rename)。
// remaining が空の場合はファイルを削除する。
// j.mu は呼び出し側が保持していること。
func (j *Journal) writeAllLocked(remaining []journalEntry) {
	if len(remaining) == 0 {
		if err := os.Remove(j.path); err != nil && !os.IsNotExist(err) {
			slog.Warn("journal: failed to remove empty journal", "path", j.path, "err", err)
		}
		return
	}
	tmpPath := j.path + fmt.Sprintf(".%d.tmp", os.Getpid())
	f, err := os.OpenFile(tmpPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o600)
	if err != nil {
		slog.Warn("journal: failed to create tmp file for rewrite", "tmp", tmpPath, "err", err)
		return
	}
	defer func() {
		if f != nil {
			f.Close()
			os.Remove(tmpPath)
		}
	}()
	w := bufio.NewWriter(f)
	for _, e := range remaining {
		data, _ := json.Marshal(e)
		fmt.Fprintf(w, "%s\n", data)
	}
	if err := w.Flush(); err != nil {
		slog.Warn("journal: flush error on rewrite", "err", err)
		return
	}
	if err := f.Sync(); err != nil {
		slog.Warn("journal: fsync error on rewrite", "err", err)
		return
	}
	f.Close()
	f = nil // suppress defer cleanup
	if err := os.Rename(tmpPath, j.path); err != nil {
		slog.Warn("journal: atomic rename failed", "tmp", tmpPath, "dst", j.path, "err", err)
	}
}
